# Heart Attack Prediction
I will analyze the Heart Attack data set from Kaggle. I will compare the performance of different models: Logistic Regression and Random Forest.



In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

In [2]:
# Loading data
data = pd.read_csv('heart.csv')

# Data analysis
Data consists of 13 features and 1 column of labels (output). Dimension of the data is: (303,14). The features are:

Age : Age of the patient - categorical

Sex : Sex of the patient - categorical

exang: exercise induced angina (1 = yes; 0 = no) - categorical

ca: number of major vessels (0-3) - categorical

cp : Chest Pain type chest pain type - categorical

    Value 1: typical angina
    Value 2: atypical angina
    Value 3: non-anginal pain
    Value 4: asymptomatic

trtbps : resting blood pressure (in mm Hg) - continuous

chol : cholestoral in mg/dl fetched via BMI sensor - continuous

fbs : (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false) - categorical

rest_ecg : resting electrocardiographic results - categorical

    Value 0: normal
    Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
    Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria

thalach : maximum heart rate achieved - continuous



output : 0 = less chance of heart attack 1= more chance of heart attack - categorical




In [3]:
# Analysing data
print(data.head())
print(data.shape)

   age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1  
(303, 14)


The data does not have any missing values.

Columns age, trtbps, chol, thalachh, oldpeak are numerical columns. The rest are categorical columns.

In [4]:
# Missing values check
print(data.isnull().sum())

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64


Before creating a model, I will `normalize` and split the data to train and test sets. I will use 80% of the data for training and 20% for testing.



In [5]:
data['age'] = MinMaxScaler().fit_transform(data['age'].values.reshape(-1,1)) # Normalizing age column
data['trtbps'] = MinMaxScaler().fit_transform(data['trtbps'].values.reshape(-1,1)) # Normalizing trtbps column
data['chol'] = MinMaxScaler().fit_transform(data['chol'].values.reshape(-1,1)) # Normalizing chol column
data['thalachh'] = MinMaxScaler().fit_transform(data['thalachh'].values.reshape(-1,1)) # Normalizing thalachh column
data['oldpeak'] = MinMaxScaler().fit_transform(data['oldpeak'].values.reshape(-1,1)) # Normalizing oldpeak column


In [6]:
data

age  sex  cp    trtbps      chol  fbs  restecg  thalachh  exng  \
0    0.708333    1   3  0.481132  0.244292    1        0  0.603053     0   
1    0.166667    1   2  0.339623  0.283105    0        1  0.885496     0   
2    0.250000    0   1  0.339623  0.178082    0        0  0.770992     0   
3    0.562500    1   1  0.245283  0.251142    0        1  0.816794     0   
4    0.583333    0   0  0.245283  0.520548    0        1  0.702290     1   
..        ...  ...  ..       ...       ...  ...      ...       ...   ...   
298  0.583333    0   0  0.433962  0.262557    0        1  0.396947     1   
299  0.333333    1   3  0.150943  0.315068    0        1  0.465649     0   
300  0.812500    1   0  0.471698  0.152968    1        1  0.534351     0   
301  0.583333    1   0  0.339623  0.011416    0        1  0.335878     1   
302  0.583333    0   1  0.339623  0.251142    0        0  0.786260     0   

      oldpeak  slp  caa  thall  output  
0    0.370968    0    0      1       1  
1    0.564516    0    0      2       1  
2    0.225806    2    0      2       1  
3    0.129032    2    0      2       1  
4    0.096774    2    0      2       1  
..        ...  ...  ...    ...     ...  
298  0.032258    1    0      3       0  
299  0.193548    1    0      3       0  
300  0.548387    1    2      3       0  
301  0.193548    1    1      3       0  
302  0.000000    1    1      2       0  

[303 rows x 14 columns]

In [7]:
y = data['output'] # Target variable
X = data.drop(['output'], axis=1) # Features

# Train test split

I will use `train_test_split` from `sklearn.model_selection` to split the data to train and test sets. I will use 80% of the data for training and 20% for testing. I will use `random_state=42` to make the results reproducible.

In [8]:
# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
X_train.shape

(242, 13)

# Logistic Regression

The first model I will use is `LogisticRegression` from `sklearn.linear_model`. I will use default parameters.

In [10]:
# Logistic Regression
log_reg = LogisticRegression() # Creating model with default parameters

In [11]:
log_reg.fit(X_train, y_train) # Fitting model

LogisticRegression()

# Model Performance
Our logistic regression model achieved an accuracy of ~84% on the test set. The model is not overfitting, because the accuracy on the test and training set are similar. We can interpret achieved accuracy as follows: for 100 patients on average 16 of them would get the wrong diagnosis.

In [12]:
print(f'Average accuracy on test set: {log_reg.score(X_test, y_test)}') # Accuracy on test set
print(f'Average accuracy on train set: {log_reg.score(X_train, y_train)}') # Accuracy on train set

Average accuracy on test set: 0.8360655737704918
Average accuracy on train set: 0.8553719008264463


The precision of the model is ~0.86. This means that for 100 patients that the model predicted to have heart disease, 86 of them actually have heart disease.

The sensitivity of the model is ~0.81. This means that for 100 patients that actually have heart disease, 81 of them are predicted to have heart disease by the model.

The F1 score of the model is ~0.83. This means that the model is good at predicting heart disease. The F1 score is the harmonic mean of precision and sensitivity.

In [13]:
# Model performance
print(f'Precision: {sklearn.metrics.precision_score(y_test, log_reg.predict(X_test))}')
print(f'Sensitivity: {sklearn.metrics.recall_score(y_test, log_reg.predict(X_test))}')
print(f'F1 score: {sklearn.metrics.f1_score(y_test, log_reg.predict(X_test))}')

Precision: 0.8666666666666667
Sensitivity: 0.8125
F1 score: 0.8387096774193549


To make the coefficients easy to interpret we need to exponentiate them. The exponentiated coefficients are as follows:

In [14]:
np.exp(log_reg.coef_) # Coefficients of the model

array([[0.60125786, 0.27105444, 2.09489142, 0.45127838, 0.65790576,
        1.20779825, 1.54097501, 2.55902379, 0.32690452, 0.20597206,
        2.51901646, 0.45119277, 0.42401766]])

# Coefficients interpretation
The general interpretation of the coefficients is as follows: For every `one-unit` increase in the variable, the odds that the patient has `more` chance of heart attack are the `coefficient times` as large as the odds that the patient has a lower chance of heart attack when all other variables are `held constant.

Values of the `coefficients` are all `similar` and in `reasonable range`, suggesting that all features are `important` for the model, therefore we can use `all of them.

age ~ 0.6

For every one-unit increase in age, the odds that the patient has more chance of heart attack are 0.6 times as large as the odds that the patient has a lower chance of heart attack when all other variables are held constant. This means that the `older` the patient is, the `more chance` of a heart attack he has.

sex ~ 0.27 

The odds of Males having more chance of heart attack are 0.27 times as large as the odds of Female ceteris paribus.

cp ~ 2.1

As variable CP increases by one unit, the odds that the patient has more chance of heart attack are 2.1 times as large as the odds that the patient has a lower chance of heart attack when all other variables are held constant. 

trtbps ~ 0.45

For every one-unit increase in trtbps, the odds that the patient has more chance of heart attack are 0.45 times as large as the odds that the patient has a lower chance of heart attack when all other variables are held constant.

chol ~ 0.66

For every one-unit increase in chol, the odds that the patient has more chance of heart attack are 0.66 times as large as the odds that the patient has a lower chance of heart attack when all other variables are held constant.

fbs ~ 1.2

The odds of having more chance of heart attack for patients with fbs > 120 mg/dl are 1.2 times as large as the odds of having a lower chance of heart attack ceteris paribus.

restecg ~ 1.54

The odds of having more chance of heart attack for patients with restecg = 1 are 1.54 times as large as the odds of having a lower chance of heart attack ceteris paribus.

thalachh ~ 2.56

For every one-unit increase in thalachh, the odds that the patient has more chance of heart attack are 2.56 times as large as the odds that the patient has a lower chance of heart attack when all other variables are held constant.

exng ~ 0.32

The odds of having more chance of heart attack for patients with exercise-induced angina are 0.32 times as large as the odds of patients without induced angina ceteris paribus.

oldpeak ~ 0.20

For every one-unit increase in oldpeak, the odds that the patient has more chance of heart attack are 0.20 times as large as the odds the patient has a lower chance of heart attack when all other variables are held constant.

etc.


`Comment: I used the word 'unit' because the data is normalized.`



# Random Forest
The next model I will use is `RandomForestClassifier` from `sklearn.ensemble`. I will use default parameters.

In [15]:
# RandomForestClassifier
forest = RandomForestClassifier() # Creating model with default parameters

In [16]:
forest.fit(X_train, y_train) # Fitting model

RandomForestClassifier()

There is a high chance that the model is `overfitting` because the accuracy on the training set is perfect, while the accuracy on test set is ~0.85. 

In [17]:
print(f'Average accuracy on test set: {forest.score(X_test, y_test)}') # Accuracy on test set
print(f'Average accuracy on train set: {forest.score(X_train, y_train)}') # Accuracy on train set

Average accuracy on test set: 0.8524590163934426
Average accuracy on train set: 1.0


We can interpret achieved `accuracy` (on test set) as follows: for 100 patients on average 15 of them would get wrong diagnosis.

`Precision` of the model is ~0.87. This means that for 100 patients that the model predicted to have heart disease, 87 of them actually have heart disease.

`Sensitivity` of the model is ~0.84. This means that for 100 patients that actually have heart disease, 84 of them are predicted to have heart disease by the model.

`F1 score` of the model is ~0.86. This means that the model is good at predicting heart disease. The F1 score is the harmonic mean of precision and sensitivity.

In [18]:
print(f'Precision: {sklearn.metrics.precision_score(y_test, forest.predict(X_test))}')
print(f'Sensitivity: {sklearn.metrics.recall_score(y_test, forest.predict(X_test))}')
print(f'F1 score: {sklearn.metrics.f1_score(y_test, forest.predict(X_test))}')

Precision: 0.8709677419354839
Sensitivity: 0.84375
F1 score: 0.8571428571428571


# Checking feature importance

The score for feature importance is the `mean decrease in impurity` (MDI) across all trees in the forest. The higher the score, the more important the feature. The score for each feature is as follows:

In [19]:
forest.feature_importances_

array([0.09379158, 0.04658748, 0.11263448, 0.06518745, 0.08287688,
       0.01207317, 0.01609314, 0.10488445, 0.08243148, 0.13642377,
       0.04464606, 0.12251505, 0.07985502])

The sixth score corresponding to feature `fbs` is the lowest, which means that the feature is the least important. The rest of the features are important. I will create a new model with only important features.

In [20]:
X_train = X_train.drop('fbs', axis = 1) # Dropping fbs column
X_test = X_test.drop('fbs', axis = 1) # Dropping fbs column

In [21]:
forest.fit(X_train, y_train) # Fitting model

RandomForestClassifier()

Interestingly, model with only important features achieved lower `accuracy` of ~0.84 on test set compared to the one with all features.

In [22]:
print(f'Average accuracy on test set: {forest.score(X_test, y_test)}') # Accuracy on test set
print(f'Average accuracy on train set: {forest.score(X_train, y_train)}') # Accuracy on train set

Average accuracy on test set: 0.8524590163934426
Average accuracy on train set: 1.0


The rest of the metrics are also lower.

In [23]:
print(f'Precision: {sklearn.metrics.precision_score(y_test, forest.predict(X_test))}')
print(f'Sensitivity: {sklearn.metrics.recall_score(y_test, forest.predict(X_test))}')
print(f'F1 score: {sklearn.metrics.f1_score(y_test, forest.predict(X_test))}')

Precision: 0.8484848484848485
Sensitivity: 0.875
F1 score: 0.8615384615384615


# Summary
In this notebook, I analyzed the Heart Attack dataset. I created a logistic regression model and a random forest model. I used default parameters for both models. I used accuracy, precision, sensitivity, and F1 score to evaluate the models. Based on the values of coefficients of logistic regression I concluded that all of the features are important. I also checked the feature importance for the random forest model. I created a new model with only important features. The model with all features achieved better results than the model with only important features. Overall, the random forest model achieved better results than the logistic regression model. It is important to note that though the logistic regression model achieved worse results, it is easier to interpret the coefficients of the logistic regression model and therefore it is easier to explain the model to the non-technical audience.